### PR TITLE
Fix Context.keymap_new_from_names()

### LINKED ISSUE #5: "TypeError: initializer for ctype 'char *' must be a cdata pointer, not bytes" when creating keymap
Hi. I tried to create a keymap with `keymap_new_from_names` but got error "TypeError: initializer for ctype 'char *' must be a cdata pointer, not bytes".  Am I wrong to pass a string argument?

```python3
>>> from xkbcommon import xkb
>>> ctx = xkb.Context()
>>> ctx.keymap_new_from_names(layout="gb")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/matt/.local/lib/python3.7/site-packages/xkbcommon/xkb.py", line 297, in keymap_new_from_names
    names.layout = layout.encode()
TypeError: initializer for ctype 'char *' must be a cdata pointer, not bytes
```

### PR DESCRIPTION
**Pull request by sde1000**
It appears this had never been tested other than with null arguments!

@flacjacket — please let me know whether this looks sane.

@hickford — does this fix your problem?

### CODE REVIEW

File: xkbcommon/xkb.py [hunk 1]
```diff
@@ -288,17 +289,14 @@ def keymap_new_from_names(self, rules=None, model=None, layout=None,
         Returns a Keymap compiled according to the RMLVO names, or
         None if the compilation failed.
         """
+        wkd = weakref.WeakKeyDictionary()
```
Comment (flacjacket) [inline xkbcommon/xkb.py]:
```suggestion
```

File: xkbcommon/xkb.py [hunk 2]
```diff
@@ -288,17 +289,14 @@ def keymap_new_from_names(self, rules=None, model=None, layout=None,
...
-            names.rules = rules.encode()
-        if model:
-            names.model = model.encode()
-        if layout:
-            names.layout = layout.encode()
-        if variant:
-            names.variant = variant.encode()
-        if options:
-            names.options = options.encode()
+        wkd[names] = []
```
Comment (flacjacket) [inline xkbcommon/xkb.py]:
```suggestion
        keep_alive = []
```

File: xkbcommon/xkb.py [hunk 3]
```diff
@@ -288,17 +289,14 @@ def keymap_new_from_names(self, rules=None, model=None, layout=None,
...
+        for x in ("rules", "model", "layout", "variant", "options"):
+            if locals()[x]:
+                c = ffi.new("char[]", locals()[x].encode())
+                setattr(names, x, c)
+                wkd[names].append(c)
```
Comment (flacjacket) [inline xkbcommon/xkb.py]:
```suggestion
                keep_alive.append(c)
```

**Review by flacjacket — COMMENTED** [review]
Thanks for finding this, you're right I only tested this without passing in parameters! In general, this looks good, just a minor note.

I think in this case, I don't think the weak key dictionary is necessary. The weak key dictionary is useful if you were needing to return an object that has ownership of some ffi allocated objects, for example, if this needed to return the `names` variable, then it would make sense to use the weak key dictionary to keep the `ffi.new(...)` objects around as long as the returned object is kept alive. In that case, the `wkd` instance would need to be a global (similar to the example in the ffi docs that you linked). However, here, the `names` object is only scoped to this function, so it is sufficient to just create a list of objects that need to be kept alive.

I've added a set of suggested changes that simplify this a bit by removing the weak key dictionary and should still work as expected.

Since the objects only need to be kept alive locally, another possible solution would be something like:
```
if rules:
    names.rules = rules = ffi.new("char []", rules.encode())
```
This should keep the `ffi.new()` allocated object in `rules`. I guess that depends what you stylistically prefer between the for loop over the variables that are allocated in `locals()` vs the chain of if statements, but I'd leave that up to you.

**Comment by sde1000** [comment]
Hmm. I can't find anything in the python execution model that guarantees that `keep_alive` would never be garbage collected until the end of the function (although I'm sure that in the current implementations it wouldn't be). A python interpreter that performed liveness analysis, for example, might garbage collect the object referred to by `keep_alive` immediately before the call to `lib.xkb_keymap_new_from_names()` since there are no further references to it in the function.

This would never normally matter, but cffi is exposing more of python's implementation details than one regularly experiences! I prefer the `WeakKeyDictionary()` approach because it makes it clear that `names` depends on the ffi pointer objects.

**Comment by flacjacket** [comment]
Couldn't you say the same thing about the `wkd` variable that it could possibly be garbage collected if the interpreter were to reason about it not being referenced later in the function? I would think that would also remove the objects that need to be kept alive. That said, I do see your point about the weak key dictionary encoding the semantics of `name` keeping the objects alive, but a comment may be helpful in in explaining what is going on even if it just links to the cffi documentation.

Reading [this](https://docs.python.org/3/reference/datamodel.html#objects-values-and-types), the `keep_alive` object and its contained objects should be kept alive as long as the variable is reachable, which I would take to be to the end of the function scope, rather than as long as the variable is used, but I don't see a definition there for "reachable". But, it looks like from the note on the CPython implementation detail, that the CPython interpreter is as aggressive as the data model allows in removing unreachable variables, at least in the absence of cyclic references.

**Comment by sde1000** [comment]
Good point!

How about using the `keep_alive` list, and then explicitly unbinding both it and `names` after the call to `lib.xkb_keymap_new_from_names()` with a suitable comment about cffi memory management? At this point I think it's just a matter of style and making sure nobody (including me!) breaks it in the future.

**Review by flacjacket — APPROVED** [review]
This looks great!

**Pull request merged**